ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/parsers/TomlParser.kt

```diff
@@ -135,52 +128,91 @@ public value class TomlParser(private val config: TomlInputConfig) {
 
                 comments.clear()
             }
+            index++
         }
         return tomlFileHead
     }
 
     /**
-     * @return true if string is a first line of multiline value declaration
+     * @param collectTo append all multi-lines to this argument
+     * @return index at the end of multiline
      */
-    private fun String.isStartOfMultilineValue(): Boolean {
+    private fun collectMultiline(
+        mutableTomlLines: List<String>,
+        collectTo: StringBuilder,
+        startIndex: Int,
+        multilineType: MultilineType,
+        comments: MutableList<String>
+    ): Int {
+        var index = startIndex
+        var lineNo = index + 1
+        var line: String
+
+        while (index < mutableTomlLines.size) {
+            line = mutableTomlLines[index]
+            if (multilineType == MultilineType.ARRAY) {
+                collectTo.append(line.takeBeforeComment(config.allowEscapedQuotesInLiteralStrings))
+                comments += line.trimComment(config.allowEscapedQuotesInLiteralStrings)
+            } else {
+                // we can't have comments inside a multi-line basic/literal string
+                collectTo.append(line)
+            }
+
+            val isFirstLine = index == startIndex
+            if (!isFirstLine && line.isEndOfMultilineValue(multilineType, lineNo)) {
+                break
+            }
+            // append new line to collect string as is
+            collectTo.append("\n")
+            index++
+            lineNo++
+        }
+
+        if (index >= mutableTomlLines.size) {
+            throw ParseException(
+                "Expected (${multilineType.closingSymbols}) in the end of ${multilineType.name}",
+                startIndex + 1
+            )
+        }
+        return index
+    }
+
+    private fun String.getMultilineType(): MultilineType {
         val line = this.takeBeforeComment(config.allowEscapedQuotesInLiteralStrings)
         val firstEqualsSign = line.indexOfFirst { it == '=' }
         if (firstEqualsSign == -1) {
-            return false
+            return MultilineType.NOT_A_MULTILINE
         }
         val value = line.substring(firstEqualsSign + 1).trim()
 
-        return value.startsWith("[") &&
-                value.endsWith("]").not()
+        if (value.startsWith("[") && !value.endsWith("]")) {
```

Comment:
the only stylistic comment I have here - is usage of `return when { 

}
`

It usually looks much better with a huge number of `if statements`